src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -99,20 +115,37 @@ public VisitResult(TypeWithState resultType, TypeSymbolWithAnnotations lValueTyp
         /// </summary>
         private readonly ArrayBuilder<(BoundReturnStatement, TypeSymbolWithAnnotations)> _returnTypesOpt;
 
-        /// <summary>
-        /// An optional callback for callers to receive notification of the inferred type and nullability
-        /// of each expression in the method. Since the walker may require multiple passes, the callback
-        /// may be invoked multiple times for a single expression, potentially with different nullability
-        /// each time. The last call for each expression will include the final inferred type and nullability.
-        /// </summary>
-        private readonly Action<BoundExpression, TypeSymbolWithAnnotations> _callbackOpt;
-
         /// <summary>
         /// Invalid type, used only to catch Visit methods that do not set
         /// _result.Type. See VisitExpressionWithoutStackGuard.
         /// </summary>
         private static readonly TypeWithState _invalidType = new TypeWithState(ErrorTypeSymbol.UnknownResultType, NullableFlowState.NotNull);
 
+        /// <summary>
+        /// Contains the map of expressions to inferred nullabilities and types used by the optional rewriter phase of the
+        /// compiler.
+        /// </summary>
+        private readonly PooledDictionary<BoundExpression, (NullabilityInfo Info, TypeSymbol Type)> _analyzedNullabilityMapOpt;
+
+        // PROTOTYPE(nullable-api): remove this when all expression are supported
+        private bool _disableNullabilityAnalysis;
+
+#if DEBUG
+        /// <summary>
+        /// Contains the expressions that should not be inserted into <see cref="_analyzedNullabilityMapOpt"/>.
+        /// </summary>
+        private static readonly ImmutableArray<BoundKind> s_skippedExpressions = ImmutableArray.Create(BoundKind.ArrayInitialization,
+            BoundKind.ObjectInitializerExpression,
+            BoundKind.CollectionElementInitializer,
+            BoundKind.DynamicCollectionElementInitializer);
+#endif
+
+        /// <summary>
+        /// Contains the expression currently being analyzed. Used by the various SetResult helpers to set the correct nullability
+        /// in <see cref="_analyzedNullabilityMapOpt"/>.
+        /// </summary>
+        private BoundExpression _currentExpression;
```

Comment:
>_currentExpression [](start = 32, length = 18)

Can this be passed explicitly to the methods that use it rather than maintaining a field?

Reply:
Unfortunately, not without turning the ResultType/RValueType properties into methods that pass along an expression. I can do either, but I chose to keep the existing style.

---
In reply to: [265682591](https://github.com/dotnet/roslyn/pull/34017#discussion_r265682591) [](ancestors = 265682591)

Reply:
We should limit the mutable state, particularly cases where two fields (in this case, `_visitResult` and `_currentExpression`) can be out of sync.

---
In reply to: [265696230](https://github.com/dotnet/roslyn/pull/34017#discussion_r265696230) [](ancestors = 265696230,265682591)